### DISCUSSION TITLE
VILA-M3 Demo Feedback

### DISCUSSION THREAD

**Discussion by holgerroth** [Project-MONAI/VLM-Radiology-Agent-Framework#35] [General]
Please provide any feedback you might have regarding using the [VILA-M3 demo](https://vila-m3-demo.monai.ngc.nvidia.com).

**Comment by amitjc** [comment]
Dear Experts,

Congratulations on the brilliant project.

Looking forward to following features:
1. Provision to run the inference on user data
2. Support for DICOM
3. User feedback for model fine-tuning if there's any inaccuracy in the AI assistant prediction.

Thanks and Regards,
Amit.

**Comment by mingxin-zheng** [comment]
Hi @amitjc , For the feedback, I agree with 1 and 2. For 3, I'm unclear what forms of user feedback may be. Can you give an example how that should work? 

Thank you!

**Reply by amitjc** [reply]
Hello @mingxin-zheng,

Thank you for your response. 

By user feedback, I meant, that there can be provision to manually correct segmentation, if there's segmentation error made by VISTA or add/edit a finding, for eg., Pneumothorax, in case such a finding hasn't been detected by the VLM AI assistant. 

Will the feature to run inference on user data be enabled in near future?

Look forward to hearing from the team.

Best Wishes, 
Amit.

**Reply by mingxin-zheng** [reply]
That makes sense!

The online demo will probably not run inference on user data very soon, due to data security concerns.
But I can try to add that in the local demo first (which requires local GPU to run):
https://github.com/Project-MONAI/VLM/blob/main/m3/demo/gradio_m3.py

**Reply by holgerroth** [reply]
We could add some options for users to correct the model output and record their desired result. That collected feedback could be used for future fine-tuning of VILA-M3. Might need some crowd-sourcing effort to collect enough real user feedback.

CC @finalelement @dongyang0122

**Reply by mingxin-zheng** [reply]
Just FYI @amitjc , I am adding an upload button for Nifti and 2D Images here
https://github.com/Project-MONAI/VLM/pull/59

Unfortunately due to [some Gradio limitation](https://github.com/gradio-app/gradio/issues/7515), there will be no display such as progress bar for the uploading process.

**Reply by amitjc** [reply]
@mingxin-zheng that's wonderful. Thanks a ton.

I tried the docker method with the latest code, however, couldn't find the upload button. PFA [screenshot.](https://github.com/user-attachments/assets/36f6c62e-354b-41c5-a603-165aaf5fc83b)

Can you help?

**Reply by mingxin-zheng** [reply]
Hi @amitjc , the PR hasn't been merged and so it's not in the `main` branch. You may use `docker exec` to get into the container, and checkout the development branch `mingxinz/upload_user_data_58` to try it.

```
docker exec -it <container_id> bash
cd /VLM
git checkout mingxinz/upload_user_data_58
```

**Reply by amitjc** [reply]
@mingxin-zheng Thank you for your continued support.

docker exec didn’t enable the upload button.

I executed the command after docker run. Shall I also modify the gradio_m3.py in the downloaded source file as well?

**Reply by mingxin-zheng** [reply]
Hi @amitjc , sorry I did not explain this clear enough. Can you please check this out and see if you can see the button? I also have the commands to use in the container
https://github.com/Project-MONAI/VLM/pull/59#issuecomment-2499609826

**Reply by amitjc** [reply]
That's awesome. Thanks.

However, after running an inference on one user radiograph, any subsequent upload is displayed as the prior uploaded image.

**Reply by mingxin-zheng** [reply]
Hi @amitjc , can you try `git pull` in the container to receive my latest fix here:
https://github.com/Project-MONAI/VLM/pull/59#issuecomment-2504209947

I hope this would solve the issue you met. Thanks!

**Reply by amitjc** [reply]
Hello @mingxin-zheng,
Thanks for the update. 
I used the following commands in the container:
```
cd /VLM/m3/demo
git fetch --all
git checkout mingxinz/upload_user_data_58
python gradio_m3.py --port 7860
```
This does show the n users equal to the number of images uploaded, but the image for every user is the 1st image uploaded.

**Reply by amitjc** [reply]
any updates?

**Reply by mingxin-zheng** [reply]
Hi @amitjc , sorry for my delayed reply. I took a break to end 2024.

I couldn't replicate what you mentioned above. On my side, it's working okay. The video should explain what I saw, by starting with no user-uploaded image. Then I uploaded the 1st one (text-labeled with "User 1") and the 2nd one ("User 2) subsequently.

There is a small delay for the UI to refresh around 00:00:29. Is that what confuses you?

https://github.com/user-attachments/assets/696b3048-5c77-4bc3-a12a-2deb5cb13013

**Reply by mingxin-zheng** [reply]
Hi @amitjc , please let me know if we can merge this feature:
https://github.com/Project-MONAI/VLM/pull/59

Thanks!

**Reply by amitjc** [reply]
Hello @mingxin-zheng,

Apologies for my delayed response.

I’m still struggling to upload multiple X-ray cases. For instance after analysing the first radiograph, uploading next image doesn’t display the new uploaded image/s, instead only earlier uploaded image is shown. I don’t know what I’m missing.

It’s definitely a desirable feature, that you have added for the local instance, to run the inference on user data.

You may consider merging the feature.

Thanks.

Best Wishes,
Amit.

**Reply by mingxin-zheng** [reply]
Thanks @amitjc . I created [this issue](https://github.com/Project-MONAI/VLM/issues/61) to track.

I will merge my PR first and investigate if I can reproduce the error you met again. 

> Hello @mingxin-zheng,
> 
> Apologies for my delayed response.
> 
> I’m still struggling to upload multiple X-ray cases. For instance after analysing the first radiograph, uploading next image doesn’t display the new uploaded image/s, instead only earlier uploaded image is shown. I don’t know what I’m missing.
> 
> It’s definitely a desirable feature, that you have added for the local instance, to run the inference on user data.
> 
> You may consider merging the feature.
> 
> Thanks.
> 
> Best Wishes, Amit.

**Comment by michelodu** [comment]
Hello, 
I'm a prof, previously with Kitware prior to coming to Old Dominion University, and I have a question regarding Vila M3. I have suggested Vila M3 to a PhD student who would like to pursue VLMs for pathology applications, which includes 2D microscopy images. This student reported not being able to get Vila M3 to work with microscopy images because of their 2D format, given that the code base emphasizes 3D. This seems to me like a trivial exercise, converting 3D to 2D, but the student claims otherwise. What would be involved in adapting Vila M3 to 2D analysis applications? 
Thanks for your support. Best wishes, 
Michel
